### ISSUE TITLE
Files does not copy correctly

### ISSUE THREAD

**Issue by OmarShawky1** [elementary/files#1784]
## What Happened
Files were copied every time with corruption.

## Expected Behavior

I was creating a bootable Ventoy USB drive (program that enables to boot from multiple ISOs without formatting) and each time I copied elementary ISO into the flash drive it refuses to boot and gives me this [booting error ](https://askubuntu.com/questions/5996/unable-to-mount-dev-loop0-during-install). one suggested to perform a checksum on the ISO and it turns out that it was corrupted and i had to open windows and copy elementary ISO and it worked this time!

## Steps to Reproduce
1. [Download ventoy-1.0.49-linux.tar.gz](https://www.ventoy.net/en/download.html) 
2. Run Ventoy web service, [(Useful tutorial for installing Ventoy in linux](https://linoxide.com/how-to-install-ventoy-on-ubuntu/)
3. Now mount the Ventoy USB stick and simply compy elementary ISO and try to boot from it!

### Logs
(in the boot error link)

##Extra Information
1. elementary.files seemed to copy files very fast compared to windows, specifically with 57 MB/s compared to 2.6 MB/s in Windows 10
2. After each time i tried to copy elementary ISO, a service called plymouth prevented the system from poweroff and it remained till eternity until i physically remove the USB from the laptop even if i unmounted it before rebooting, the issue remained the same.
3. this issue did not appear when i copied Windows.ISO as the windows worked with no issues at all!

## Platform Information
OS: elementary OS 6 Odin x86_64 (20210810 built)
Hardware: Lenovo Legion Y530

**Comment by jeremypw** [comment]
I am not sure it is the cause of your issue, but when Files copies files to a slow drive, they get cached in memory by the underlying virtual filesystem (gvfs) - hence the apparently fast transfer rate.  So when Files finishes the copy, data is still being transferred to the drive by the system in the background.  This is probably what is inhibiting the poweroff.  Removing the drive in the middle of a transfer is likely to corrupt the data. From your quoted transfer rates it looks like you will have to wait up to 20x the time it took Files to finish the transfer for all the data to actually be transferred to the drive.  You could try monitoring the available space on the drive to check whether it is still changing.

**Comment by jeremypw** [comment]
Issue https://github.com/elementary/files/issues/128 looks similar.

**Comment by jeremypw** [comment]
Issue #1748 is also relevant.

**Comment by abodedis** [comment]
Agree with everything said here and in the linked issues! Wish there was a way to escalate this to the highest priority, I learned about this the hard way with **so many** corrupted transfers before I figured this out.

**Comment by jeremypw** [comment]
*Always* unmount a removable drive before unplugging it.  In theory, if data is still is still being transferred to it the system (not Files) should throw up a warning dialog.  Also some USB sticks have a warning light that flashes while it is still being written too. But there is little that can be done to enforce this.  Maybe a warning message of some kind can be shown when the transfer ends as far as Files is concerned not to unplug the drive before unmounting?

The alternative would be to lock up Files completely during a long file transfer - which would probably trigger issues raised from others.

**Comment by OmarShawky1** [comment]
> _Always_ unmount a removable drive before unplugging it. In theory, if data is still is still being transferred to it the system (not Files) should throw up a warning dialog. Also some USB sticks have a warning light that flashes while it is still being written too. But there is little that can be done to enforce this. Maybe a warning message of some kind can be shown when the transfer ends as far as Files is concerned not to unplug the drive before unmounting?
> 
> The alternative would be to lock up Files completely during a long file transfer - which would probably trigger issues raised from others.

Please understand my issue and re-read it. Yes ... I did entirely unmount it and even gave it couple of minutes before i try to turnoff my laptop

**Comment by abodedis** [comment]
> > _Always_ unmount a removable drive before unplugging it. In theory, if data is still is still being transferred to it the system (not Files) should throw up a warning dialog. Also some USB sticks have a warning light that flashes while it is still being written too. But there is little that can be done to enforce this. Maybe a warning message of some kind can be shown when the transfer ends as far as Files is concerned not to unplug the drive before unmounting?
> > The alternative would be to lock up Files completely during a long file transfer - which would probably trigger issues raised from others.
> 
> Please understand my issue and re-read it. Yes ... I did entirely unmount it and even gave it couple of minutes before i try to turnoff my laptop

Can you try coping a lot of normal files to a USB drive (not burning an ISO), waiting until the process is complete, unmount the USB drive, remount the USB drive, and verify if the data on the USB drive is corrupted? If not, then the issue may be the ISO burning process.

**Comment by OmarShawky1** [comment]
> > > _Always_ unmount a removable drive before unplugging it. In theory, if data is still is still being transferred to it the system (not Files) should throw up a warning dialog. Also some USB sticks have a warning light that flashes while it is still being written too. But there is little that can be done to enforce this. Maybe a warning message of some kind can be shown when the transfer ends as far as Files is concerned not to unplug the drive before unmounting?
> > > The alternative would be to lock up Files completely during a long file transfer - which would probably trigger issues raised from others.
> > 
> > 
> > Please understand my issue and re-read it. Yes ... I did entirely unmount it and even gave it couple of minutes before i try to turnoff my laptop
> 
> Can you try coping a lot of normal files to a USB drive (not burning an ISO), waiting until the process is complete, unmount the USB drive, remount the USB drive, and verify if the data on the USB drive is corrupted? If not, then the issue may be the ISO burning process.

1. That actually did not happen when i copied the Windows.ISO
2. This happened only when i copied elementary.ISO
3. This is not a burning process, it is just a copying one.
4. This is not the first occurrence that this issue happened; It happened a lot of times even in the old elementary files program but i was not sure what was the cause of the problem.

I hope that i made the first issue clear enough specifically that "this is not the first time for this issue; it was only the first time that i find out that elementary.files was the reason".

**Comment by jeremypw** [comment]
So it seems that unmounting does *not* trigger a warning if the mount is in use by `gvfs` (assuming that that is the cause of the problem).  I am not sure it is a Files problem if the underlying system informs it that the file operation is completed.  I do not know of a way of Files detecting that the usb is still, in fact, in use.  Maybe another more knowledgeable dev can help.

**Comment by OmarShawky1** [comment]
> So it seems that unmounting does _not_ trigger a warning if the mount is in use by `gvfs` (assuming that that is the cause of the problem). I am not sure it is a Files problem if the underlying system informs it that the file operation is completed and the unmount operation does not throw an error. I do not know of a way of Files detecting that the usb is still, in fact, in use. Maybe another more knowledgeable dev can help.

I think i provided very much enough, it is very simple to debug the issue back to gvfs, you only need the live OS working and plug in a USB and install on it Ventoy as i mentioned and try to copy elementary6.ISO, it will not take more than an hour for all the steps.

**Comment by jeremypw** [comment]
I have now downloaded Ventoy-1.0.50 and created a USB with it installed. I copied an old elementary-6.0 iso onto it using Files and can confirm that unmounting the USB after the file transfer complete notification showed did not trigger an error although the flashing light on the USB showed data was being transferred.  This continued after closing Files for about a minute and system-monitor showed two or three processes consuming a low amount of cpu. I unplugged the USB after the flashing stopped and successfully started an installation process from it.  Unplugging it while the light was still flashing would certainly have rendered it unbootable though.

If someone can suggest a way of Files detecting this ongoing data transfer I would be happy to implement e.g. blocking unmounting and/or suitable warnings in Files. Otherwise it would appear to be an upstream issue that allows the USB to be unmounted despite being in use by gvfs or whatever process is responsible for flushing the cached data to memory.

**Comment by abodedis** [comment]
> I have now downloaded Ventoy-1.0.50 and created a USB with it installed. I copied an old elementary-6.0 iso onto it using Files and can confirm that unmounting the USB after the file transfer complete notification showed did not trigger an error although the flashing light on the USB showed data was being transferred. This continued after closing Files for about a minute and system-monitor showed two or three processes consuming a low amount of cpu. I unplugged the USB after the flashing stopped and successfully started an installation process from it. Unplugging it while the light was still flashing would certainly have rendered it unbootable though.
> 
> If someone can suggest a way of Files detecting this ongoing data transfer I would be happy to implement e.g. blocking unmounting and/or suitable warnings in Files. Otherwise it would appear to be an upstream issue that allows the USB to be unmounted despite being in use by gvfs or whatever process is responsible for flushing the cached data to memory.

My USB drive does not have a flashing light, so there is no reliable way to tell when the transfer is done without looking at system monitors.  

A "normal" user may (at best) hit unmount or (more likely) just pull out the USB drive when the progress indicator is done, when in fact the progress is not done.  

As I've stated above, this seems to be a major usability issue and should be highly prioritized if possible.  If I hadn't discovered the workaround (which is to wait), I would not still use elementary OS.  Thanks for looking into this.  This will certainly help a lot of beginners.

**Comment by abodedis** [comment]
Recommendations:
1) If any file transfer is ongoing (to any drive, not just USB) and the Eject/unmount button is pressed, display an error message like "A file transfer is in progress. Please wait, or cancel the file transfer before Ejecting."

2) The progress indicator should be accurate.  100% means **done**, not "the data has been cached."  So in the case of a USB drive, expect that a user will not hit unmount and just pull out the USB drive promptly after seeing 100%, and that the data will be copied with no corruption.  Therefore, cache should already be flushed and the drive ready to be pulled out when the progress is at 100%.  If there is no way to track the progress of the cache flushing, an acceptable, not very elegant, but "better-than-nothing" solution may be to "hang" the progress indicator at 95%.  The user will then know to wait a reasonable time.  Have a spinner keep spinning.  However, the hang should not be too long or else the user will think something crashed. Then when the cache is confirmed to be flushed and the file operation has been confirmed to be complete, display 100%.

**Comment by jeremypw** [comment]
I do not disagree with the above suggestions but they still rely on Files knowing the state of the background cache, over which it has no control (afaik).  I'll look into it again but last time I came up with nothing :disappointed:  Another suggestion is to enforce direct writing to external ejectable drives, by-passing the cache, but this may need to be done at a system level.

**Comment by abodedis** [comment]
Yes, direct writing may work.

Is there any useful information that can be derived/utilized by Files, such as "size of the pending file transfer", "cache size", "estimated cache transfer rate", etc., which can help provide a more accurate estimate for the progress bar status? If progress can be accurately estimated in this way, then there's no need to worry that the "95%" hang will be too long, as the estimate would have made the progress bar move slower and more accurately.

**Comment by abodedis** [comment]
For example, if a small "test file" can be written to the drive as a precursor, for the purposes of estimating the transfer rate of the cache, and if you believe such to be a reliable metric, then the problem is essentially solved.  Of course one wouldn't really need to write the precursor files if the file transfer is trivial, such as < 50MB because those can be expected to complete quickly.

**Comment by jeremypw** [comment]
It seems from https://www.windowslatest.com/2020/09/13/windows-10-usb-quick-removal-better-performance/ that Windows 10 does disable write caching for removable drives. This is an operating system responsibility not an app one, I think as other apps could also transfer large files onto a USB stick.  The Files app is entirely reliant on `gvfs` for transfer metrics and those returned refer to the amount of data transferred into the cache unfortunately.  I agree this is unsatisfactory.

**Comment by abodedis** [comment]
Is it proper or conventional to implement direct writing to removable storage in a linux distro, at a system level, or is this one those things that simply isn't/shouldn't be done?

**Comment by jeremypw** [comment]
I can confirm that if you try to unmount a drive being written to *while the Files progress window is open* then you do get an dialog warning that the drive is in use by an app and giving the option to cancel the unmount or force it.  This is a system level dialog not from Files.

**Comment by jeremypw** [comment]
I should also say that even if Files/gvfs unmounts the drive, it presumably isn't unmounted in the kernel while it is still being written to.

> Is it proper or conventional to implement direct writing to removable storage in a linux distro, at a system level, or is this one those things that simply isn't/shouldn't be done?

I have to do more research to answer that.  Given the number of distros it would be surprising if there was a general rule unless it is simply not technically possible.  I need to check whether this issue occurs in other file managers/distros.

**Comment by jeremypw** [comment]
OK. Now I am getting somewhere.  It turns out that the `Gtk.MountOperation` has a signal `show_unmount_progress` that the default implementation ignores.  You have to subclass that class and handle the signal to show whether all data has been flushed to disk and inform the user accordingly.  Not sure whether the UX team will want to show a dialog, send a notification or just keep the unmount spinner spinning until the drive can be safely unmounted.

Seems illogical that the default implementation should provide a default handler for the `show_processes` signal to block the unmounting when other apps are using the drive but not handle the `show_unmount_progress` signal :shrug:

I should now be able to fix this.

**Comment by jeremypw** [comment]
I've just realised that PR #1752 seems to fix this problem anyway - and I wrote it :roll_eyes: It seems that if you use `mount.eject_with_operation` for ejectable mounts then Gtk *does* wait until the cache is flushed before signalling the operation has finished.

I noticed that a plugged in SD card is not regarded as "ejectable" and `mount.unmount_with_operation ()` is used, but in that case, as far as I could tell, the transfer progress window stays open until all the data is written so the system must have mounted the card without caching.

I could still subclass the default MountOperation to give more user feedback I guess but I will wait until #1752 is merged before making more changes.
